lib/s3routes/routesUtils.ts

```diff
@@ -445,23 +444,33 @@ function retrieveData(
                 });
                 // errors on server side with readable stream
                 readable.on('error', err => {
-                    log.error('error piping data from source');
-                    _destroyResponse();
+                    log.error('error piping data from source', { error: err });
                     return cbOnce(err);
                 });
                 currentStream = readable;
-                return readable.pipe(response, { end: false });
+                return readable.pipe(response, { end: false })
+                    .on('error', pipeErr => {
+                        log.error('error in pipe operation', { error: pipeErr });
+                        return cbOnce(pipeErr);
+                    });
             }), err => {
-            currentStream = null;
-            if (err) {
-                log.debug('abort response due to error', {
-                    // @ts-expect-error
-                    error: err.code, errMsg: err.message });
-            }
-            // call end for all cases (error/success) per node.js docs
-            // recommendation
-            response.end();
-        },
+                currentStream = null;
+                if (err) {
+                    log.error('abort response due to error', {
+                        // @ts-expect-error
+                        error: err.code, errMsg: err.message });
+                    // Only end if we haven't already destroyed the response
+                    if (!responseDestroyed) {
+                        return XMLResponseBackend.errorResponse(errors.ServiceUnavailable, response, log);
+                    }
+                }
+                if (codeOnSuccess && !response.headersSent) {
+                    response.writeHead(codeOnSuccess);
+                }
+                // call end for all cases (error/success) per node.js docs
```

Comment:
not clear to me why/how/when this is needed, quite confusing to send response only if headers not sent, but `okContentHeadersResponse` was changed to allow changing headers without setting response code...

I need to re-review this with a fresher mind, tomorrow morning

Reply:
This part is unchanged, but in short we should always end the request whatever we did, otherwise the client won't get notified. With the changes all paths are setting the header code, but even if there is none, we can still end the request.

But I agree that the whole code is hard to follow, the mix of paths in the same function is not trivial, plus the steps before, where we used to even set the headers before the actual request to the data service...